<h1 style="color:black; font-weight:bold">NBA Draft Analysis</h1>   


<h2 style="color:red; font-weight:bold">Import Libraries</h2>

In [75]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px


<h2 style="color:red; font-weight:bold">Load the 1989-2021 Draft Dataset</h2>

In [76]:
df = pd.read_csv('nbaplayersdraft.csv')

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1922 entries, 0 to 1921
Data columns (total 24 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   id                         1922 non-null   int64  
 1   year                       1922 non-null   int64  
 2   rank                       1922 non-null   int64  
 3   overall_pick               1922 non-null   int64  
 4   team                       1922 non-null   object 
 5   player                     1922 non-null   object 
 6   college                    1585 non-null   object 
 7   years_active               1669 non-null   float64
 8   games                      1669 non-null   float64
 9   minutes_played             1669 non-null   float64
 10  points                     1669 non-null   float64
 11  total_rebounds             1669 non-null   float64
 12  assists                    1669 non-null   float64
 13  field_goal_percentage      1665 non-null   float

In [78]:
df.describe().T

count         mean          std       min  \
id                         1922.0   961.500000   554.977927     1.000   
year                       1922.0  2005.317378     9.456946  1989.000   
rank                       1922.0    29.694589    16.912454     1.000   
overall_pick               1922.0    29.694589    16.912454     1.000   
years_active               1669.0     6.332534     4.656321     1.000   
games                      1669.0   348.042540   324.897567     1.000   
minutes_played             1669.0  8399.055722  9845.871529     0.000   
points                     1669.0  3580.413421  4826.142847     0.000   
total_rebounds             1669.0  1497.009587  2003.686388     0.000   
assists                    1669.0   774.300779  1284.602969     0.000   
field_goal_percentage      1665.0     0.436568     0.083846     0.000   
3_point_percentage         1545.0     0.272405     0.128339     0.000   
free_throw_percentage      1633.0     0.716825     0.118702     0.000   
average_minutes_played     1669.0    18.134032     8.707656     0.000   
points_per_game            1669.0     7.275734     4.969343     0.000   
average_total_rebounds     1669.0     3.194368     2.083895     0.000   
average_assists            1669.0     1.550749     1.488536     0.000   
win_shares                 1669.0    17.873697    27.989805    -1.700   
win_shares_per_48_minutes  1668.0     0.061691     0.094467    -1.264   
box_plus_minus             1668.0    -2.311271     4.143403   -52.000   
value_over_replacement     1669.0     4.403176    11.461729    -8.500   

                                25%       50%        75%        max  
id                          481.250   961.500   1441.750   1922.000  
year                       1997.000  2005.000   2013.000   2021.000  
rank                         15.000    30.000     44.000     60.000  
overall_pick                 15.000    30.000     44.000     60.000  
years_active                  2.000     5.000     10.000     22.000  
games                        72.000   235.000    584.000   1541.000  
minutes_played              838.000  4204.000  13246.000  52139.000  
points                      265.000  1552.000   5150.000  37062.000  
total_rebounds              128.000   656.000   2139.000  15091.000  
assists                      46.000   257.000    910.000  12091.000  
field_goal_percentage         0.404     0.435      0.474      1.000  
3_point_percentage            0.222     0.317      0.356      1.000  
free_throw_percentage         0.659     0.736      0.797      1.000  
average_minutes_played       11.000    17.700     24.800     41.100  
points_per_game               3.400     6.200     10.000     27.200  
average_total_rebounds        1.700     2.800      4.200     13.300  
average_assists               0.500     1.100      2.100      9.500  
win_shares                    0.400     5.300     24.500    249.500  
win_shares_per_48_minutes     0.030     0.069      0.104      1.442  
box_plus_minus               -3.900    -2.000     -0.300     51.100  
value_over_replacement       -0.400     0.000      4.500    142.600

<h2 style="color:red; font-weight:bold"> Check for NA Values</h2>

In [79]:
df.isna().sum()

id                             0
year                           0
rank                           0
overall_pick                   0
team                           0
player                         0
college                      337
years_active                 253
games                        253
minutes_played               253
points                       253
total_rebounds               253
assists                      253
field_goal_percentage        257
3_point_percentage           377
free_throw_percentage        289
average_minutes_played       253
points_per_game              253
average_total_rebounds       253
average_assists              253
win_shares                   253
win_shares_per_48_minutes    254
box_plus_minus               254
value_over_replacement       253
dtype: int64

<h2 style="color:red; font-weight:bold">Replacing All NA Values in Numerical Columns with 0 </h2>

In [80]:
na_cols = ['years_active', 'games', 'minutes_played', 'points', 'total_rebounds',
       'assists', 'field_goal_percentage', '3_point_percentage',
       'free_throw_percentage', 'average_minutes_played', 'points_per_game',
       'average_total_rebounds', 'average_assists', 'win_shares',
       'win_shares_per_48_minutes', 'box_plus_minus',
       'value_over_replacement']

for col in na_cols:
    df[col]= df[col].fillna(0)

### All players with missing college values fall into three categories:   
* Foreign Players   
* Entered the NBA straight out of High School   
* Entered the NBA straight out of the G-League    

With 337 total players that fall into these categories, I created one category to represent them together: HS/G-Lea/Foreign

In [81]:
df['college']= df['college'].fillna("HS/G-Lea/Foreign")

In [82]:
df.isna().sum()

id                           0
year                         0
rank                         0
overall_pick                 0
team                         0
player                       0
college                      0
years_active                 0
games                        0
minutes_played               0
points                       0
total_rebounds               0
assists                      0
field_goal_percentage        0
3_point_percentage           0
free_throw_percentage        0
average_minutes_played       0
points_per_game              0
average_total_rebounds       0
average_assists              0
win_shares                   0
win_shares_per_48_minutes    0
box_plus_minus               0
value_over_replacement       0
dtype: int64

With the inclusion of team relocations and expansion teams, I made the necessary adjustments to the team abbreviations in the NBA draft picks dataset. This step aims to ensure the accuracy and consistency of the team abbreviations, aligning them with the current state of the NBA teams.

In [83]:
condition = (df['team'] == 'CHH') & (df['year'] == 2014)
df.loc[condition, 'team'] = 'CHO'
team_replace = {"SEA" : "OKC", "VAN" : "MEM", "NJN" : "BRK", "WSB" : "WAS",
               "CHA" : "CHO", "NOK" : "NOP", "NOH" : "NOP", "CHH" : "NOP"}
df['team'] = df['team'].replace(team_replace)

# Exploratory Data Analysis
***
<h2 style="color:red; font-weight:bold">EDA for each NBA Team </h2>
<h3 style="color:blue; font-weight:bold">Points Per Game by Team and Overall Pick</h3>

In [84]:
sorted_df = df.sort_values(by = "team", ascending = True)

fig = px.scatter(sorted_df, x = "overall_pick", y = "points_per_game",
                    range_x = (0, 60), range_y = (0, 30),
                    hover_data = ["college"],
                    hover_name="player",
                    animation_frame = "team", range_color = (0, 25),
                    color = "points_per_game", color_continuous_scale = "YlOrRd")
                    #title = "Points Per Game Overall Pick 1-60 Animation by Year and Games")

fig.update_traces(marker = dict(size = 5)) # scaling down the markers
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

In [85]:
df.groupby(by="team")["points"].agg(["mean","min","max"])

mean  min      max
team                           
ATL   2430.912500  0.0  20894.0
BOS   3190.853333  0.0  26397.0
BRK   2760.131148  0.0  14247.0
CHI   2968.173333  0.0  20558.0
CHO   2921.358974  0.0  14414.0
CLE   4397.763636  0.0  37062.0
DAL   1935.000000  0.0  17529.0
DEN   3426.140625  0.0  28289.0
DET   3034.753846  0.0  17137.0
GSW   3856.530303  0.0  25728.0
HOU   3057.983607  0.0  17349.0
IND   2790.754386  0.0  15029.0
LAC   3098.982759  0.0  14343.0
LAL   3043.368421  0.0  14155.0
MEM   3471.673469  0.0  15116.0
MIA   3139.576923  0.0  23165.0
MIL   3635.826087  0.0  31560.0
MIN   3265.142857  0.0  26071.0
NOP   4022.327586  0.0  33643.0
NYK   2414.622642  0.0  11593.0
OKC   3577.024691  0.0  25526.0
ORL   3051.661765  0.0  28596.0
PHI   2795.900000  0.0  24368.0
PHO   2902.720000  0.0  17700.0
POR   2477.257576  0.0  19591.0
SAC   3174.904110  0.0  13647.0
SAS   2913.758065  0.0  26496.0
TOR   4164.279070  0.0  20042.0
UTA   2596.166667  0.0  14536.0
WAS   3387.338983  0.0  16159.0

<h3 style="color:blue; font-weight:bold">Total Career Games by Team and Overall Pick</h3>

In [86]:
sorted_df = df.sort_values(by = "team", ascending = True)

fig = px.scatter(sorted_df, x = "overall_pick", y = "games",
                    range_x = (0, 60), range_y = (0, 1600),
                    hover_data = ["college"],
                    hover_name="player",
                    animation_frame = "team", range_color = (0, 1600),
                    color = "games", color_continuous_scale = "YlOrRd")
                    #title = "Points Per Game Overall Pick 1-60 Animation by Year and Games")

fig.update_traces(marker = dict(size = 5)) # scaling down the markers
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

In [87]:
df.groupby(by="team")["games"].agg(["mean","min","max"])

mean  min     max
team                         
ATL   256.387500  0.0  1410.0
BOS   295.306667  0.0  1343.0
BRK   275.819672  0.0  1232.0
CHI   318.613333  0.0  1076.0
CHO   298.461538  0.0   976.0
CLE   381.454545  0.0  1366.0
DAL   219.903226  0.0  1391.0
DEN   312.906250  0.0  1260.0
DET   316.246154  0.0  1026.0
GSW   329.727273  0.0  1541.0
HOU   307.442623  0.0  1181.0
IND   311.035088  0.0  1094.0
LAC   333.017241  0.0  1160.0
LAL   324.649123  0.0  1287.0
MEM   326.816327  0.0  1023.0
MIA   281.730769  0.0  1110.0
MIL   337.043478  0.0  1522.0
MIN   305.428571  0.0  1462.0
NOP   333.396552  0.0  1346.0
NYK   269.245283  0.0  1118.0
OKC   318.024691  0.0  1335.0
ORL   282.308824  0.0  1242.0
PHI   255.822222  0.0  1223.0
PHO   263.853333  0.0  1217.0
POR   256.045455  0.0  1380.0
SAC   327.109589  0.0   997.0
SAS   283.693548  0.0  1392.0
TOR   345.744186  0.0  1083.0
UTA   304.569444  0.0  1085.0
WAS   349.559322  0.0  1208.0

<h3 style="color:blue; font-weight:bold">Total Career Rebounds by Team and Overall Pick</h3>

In [88]:
fig = px.scatter(sorted_df, x = "overall_pick", y = "average_total_rebounds",
                    range_x = (0, 60), range_y = (0, 18),
                    hover_data = ["college"],
                    hover_name="player",
                    animation_frame = "team", range_color = (0, 15),
                    color = "average_total_rebounds", color_continuous_scale = "YlOrRd")
                    #title = "Points Per Game Overall Pick 1-60 Animation by Year and Games")

fig.update_traces(marker = dict(size = 5)) # scaling down the markers
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

In [89]:
df.groupby(by="team")["total_rebounds"].agg(["mean","min","max"])

mean  min      max
team                           
ATL   1047.487500  0.0  11305.0
BOS   1118.640000  0.0   7690.0
BRK   1294.311475  0.0   8409.0
CHI   1324.000000  0.0   9040.0
CHO   1170.589744  0.0   5967.0
CLE   1596.309091  0.0  10210.0
DAL    871.161290  0.0   8725.0
DEN   1406.703125  0.0  12359.0
DET   1346.215385  0.0   9519.0
GSW   1513.060606  0.0   6854.0
HOU   1357.311475  0.0   6120.0
IND   1287.877193  0.0   8605.0
LAC   1749.775862  0.0  10467.0
LAL   1345.473684  0.0   9326.0
MEM   1426.346939  0.0   8666.0
MIA   1077.250000  0.0   7328.0
MIL   1391.768116  0.0  11489.0
MIN   1286.974026  0.0  14662.0
NOP   1466.051724  0.0   7137.0
NYK   1090.075472  0.0   7320.0
OKC   1385.209877  0.0   8834.0
ORL   1478.867647  0.0  14627.0
PHI   1174.744444  0.0   7731.0
PHO   1187.626667  0.0  10101.0
POR   1025.136364  0.0  10208.0
SAC   1411.410959  0.0   6664.0
SAS   1088.000000  0.0  15091.0
TOR   1881.186047  0.0   9513.0
UTA   1154.458333  0.0   7673.0
WAS   1351.728814  0.0   7428.0

<h3 style="color:blue; font-weight:bold">Total Career Assists by Team and Overall Pick</h3>

In [90]:
fig = px.scatter(sorted_df, x = "overall_pick", y = "average_assists",
                    range_x = (0, 60), range_y = (0, 10),
                    hover_data = ["college"],
                    hover_name="player",
                    animation_frame = "team", range_color = (0, 10),
                    color = "average_assists", color_continuous_scale = "YlOrRd")
                    #title = "Points Per Game Overall Pick 1-60 Animation by Year and Games")

fig.update_traces(marker = dict(size = 5)) # scaling down the markers
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

In [91]:
df.groupby(by="team")["assists"].agg(["mean","min","max"])

mean  min      max
team                           
ATL    577.450000  0.0   5415.0
BOS    657.746667  0.0   5636.0
BRK    564.278689  0.0   5972.0
CHI    631.680000  0.0   4245.0
CHO    665.410256  0.0   5096.0
CLE   1097.018182  0.0  10045.0
DAL    517.564516  0.0  12091.0
DEN    741.359375  0.0   4508.0
DET    589.200000  0.0   4252.0
GSW    802.060606  0.0   7095.0
HOU    582.213115  0.0   5939.0
IND    495.456140  0.0   2641.0
LAC    541.362069  0.0   3554.0
LAL    777.245614  0.0   5777.0
MEM    920.938776  0.0   6469.0
MIA    591.269231  0.0   5701.0
MIL    814.217391  0.0   6471.0
MIN    748.077922  0.0   5445.0
NOP    893.189655  0.0  10977.0
NYK    458.301887  0.0   2997.0
OKC    846.814815  0.0   8966.0
ORL    511.397059  0.0   3526.0
PHI    612.266667  0.0   5624.0
PHO    673.986667  0.0  10335.0
POR    473.121212  0.0   4726.0
SAC    694.397260  0.0   4611.0
SAS    731.983871  0.0   7036.0
TOR    713.976744  0.0   5371.0
UTA    579.708333  0.0   6819.0
WAS    745.067797  0.0   5557.0

<h3 style="color:blue; font-weight:bold">Career Span by Team and Overall Pick</h3>

In [92]:
fig = px.scatter(sorted_df, x = "overall_pick", y = "years_active",
                    range_x = (0, 60), range_y = (0, 25),
                    hover_data = ["college"],
                    hover_name="player",
                    animation_frame = "team", range_color = (0, 25),
                    color = "years_active", color_continuous_scale = "YlOrRd")
                    #title = "Points Per Game Overall Pick 1-60 Animation by Year and Games")

fig.update_traces(marker = dict(size = 5)) # scaling down the markers
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

In [93]:
df.groupby(by="team")["years_active"].agg(["mean","min","max"])

mean  min   max
team                     
ATL   4.725000  0.0  19.0
BOS   5.400000  0.0  19.0
BRK   5.032787  0.0  17.0
CHI   5.946667  0.0  17.0
CHO   5.358974  0.0  14.0
CLE   6.745455  0.0  20.0
DAL   4.338710  0.0  19.0
DEN   5.515625  0.0  19.0
DET   5.892308  0.0  18.0
GSW   5.924242  0.0  22.0
HOU   5.655738  0.0  17.0
IND   5.736842  0.0  16.0
LAC   5.931034  0.0  19.0
LAL   5.859649  0.0  18.0
MEM   5.775510  0.0  16.0
MIA   5.192308  0.0  18.0
MIL   6.304348  0.0  21.0
MIN   5.350649  0.0  21.0
NOP   5.965517  0.0  20.0
NYK   5.207547  0.0  18.0
OKC   5.518519  0.0  17.0
ORL   5.323529  0.0  19.0
PHI   4.700000  0.0  18.0
PHO   4.960000  0.0  18.0
POR   4.893939  0.0  18.0
SAC   5.917808  0.0  15.0
SAS   5.048387  0.0  19.0
TOR   5.883721  0.0  17.0
UTA   5.597222  0.0  18.0
WAS   6.050847  0.0  19.0

<h3 style="color:blue; font-weight:bold">Career Box Plus Minus by Team and Overall Pick</h3>

In [94]:
fig = px.scatter(sorted_df, x = "overall_pick", y = "box_plus_minus",
                    range_x = (0, 60), range_y = (-50, 50),
                    hover_data = ["college"],
                    hover_name="player",
                    animation_frame = "team", range_color = (-50, 50),
                    color = "box_plus_minus", color_continuous_scale = "YlOrRd")
                    #title = "Points Per Game Overall Pick 1-60 Animation by Year and Games")

fig.update_traces(marker = dict(size = 5)) # scaling down the markers
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

In [95]:
df.groupby(by="team")["box_plus_minus"].agg(["mean","min","max"])

mean   min   max
team                      
ATL  -2.508750 -16.3   6.8
BOS  -2.325333 -25.9   5.3
BRK  -2.695082 -52.0   3.0
CHI  -2.466667 -11.1   4.7
CHO  -1.728205 -21.8   6.0
CLE  -1.903636 -15.9   8.9
DAL  -2.270968 -25.5   3.8
DEN  -1.262500  -8.8   8.9
DET  -1.821538 -14.0   9.4
GSW  -2.536364 -32.4   6.5
HOU  -1.795082 -16.9   3.2
IND  -1.296491  -7.9   6.9
LAC  -1.913793  -7.0   3.4
LAL  -2.028070 -19.4   3.1
MEM  -1.755102  -9.5   3.4
MIA  -2.534615 -18.2   5.0
MIL  -1.981159 -13.0   6.0
MIN  -1.520779  -8.7   5.6
NOP  -2.025862 -20.8   7.3
NYK  -2.896226 -26.2   2.3
OKC  -1.474074 -10.1   6.8
ORL  -1.804412 -17.4   5.2
PHI  -2.090000 -29.8  17.1
PHO  -2.313333 -18.9  30.1
POR  -2.643939 -10.8   4.7
SAC  -1.789041  -9.1   2.5
SAS  -1.224194  -8.5   5.6
TOR  -0.700000 -27.7  51.1
UTA  -2.305556 -16.8   4.9
WAS  -1.996610 -10.0   2.4

<h3 style="color:blue; font-weight:bold">Value over Replacement by Team and Overall Pick</h3>

In [96]:
fig = px.scatter(sorted_df, x = "overall_pick", y = "value_over_replacement",
                    range_x = (0, 60), range_y = (-10, 60),
                    hover_data = ["college"],
                    hover_name="player",
                    animation_frame = "team",
                    color = "value_over_replacement", color_continuous_scale = "YlOrRd")
                    #title = "Points Per Game Overall Pick 1-60 Animation by Year and Games")

fig.update_traces(marker = dict(size = 5)) # scaling down the markers
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

In [97]:
df.groupby(by="team")["value_over_replacement"].agg(["mean","min","max"])

mean  min    max
team                      
ATL   2.977500 -6.4   57.5
BOS   3.662667 -5.2   65.5
BRK   3.775410 -2.4   39.2
CHI   3.037333 -4.3   38.3
CHO   2.864103 -3.2   29.8
CLE   6.665455 -5.9  142.6
DAL   1.985484 -4.0   73.5
DEN   4.390625 -2.2   44.0
DET   3.250769 -3.5   43.6
GSW   5.350000 -2.4   60.9
HOU   3.637705 -6.5   28.8
IND   3.233333 -1.7   40.4
LAC   2.465517 -8.5   32.5
LAL   4.235088 -2.9   42.3
MEM   4.971429 -4.3   41.8
MIA   3.396154 -3.4   62.8
MIL   4.315942 -3.5   84.8
MIN   4.211688 -7.0   96.9
NOP   6.718966 -4.2   93.7
NYK   2.509434 -2.7   21.8
OKC   5.692593 -4.6   76.4
ORL   4.186765 -3.7   75.5
PHI   3.445556 -3.2   49.6
PHO   3.656000 -3.1   48.2
POR   2.368182 -4.2   43.7
SAC   2.964384 -3.5   26.2
SAS   4.414516 -7.8   91.1
TOR   5.904651 -2.4   55.7
UTA   2.744444 -4.5   41.7
WAS   2.500000 -7.0   38.3

<h2 style="color:red; font-weight:bold">EDA for each College</h2>
<h3 style="color:blue; font-weight:bold">Top 5 Colleges with the Most Players Drafted</h3>

In [98]:
group_counts = df['college'].value_counts().reset_index()
group_counts.columns = ['college', 'Count']
print(group_counts.loc[1:5])


    college  Count
1  Kentucky     58
2      Duke     57
3   Arizona     43
4       UNC     43
5      UCLA     42


In [99]:
colleges_avg = df.drop(['id', 'year', 'rank', 'overall_pick', 'team'], axis=1)
colleges_avg['count'] = colleges_avg.groupby('college')['college'].transform('count')
colleges_avg = colleges_avg.groupby('college').mean().reset_index()
top25 = colleges_avg.nlargest(26, 'count')
top25 = top25.drop(67)
top25.reset_index(drop=True, inplace=True)

/var/folders/2d/rskyyr3x54v7smyxgknvdf8r0000gn/T/ipykernel_5321/4239038288.py:3: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



<h3 style="color:blue; font-weight:bold">Years Active for the Top 25 Colleges</h3>


In [100]:
fig = px.bar(top25, x='college', y='years_active', hover_data = ["count"], color='college', 
                 title = "Years Active for the Top 25 Colleges")
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

<h3 style="color:blue; font-weight:bold">Points Per Game for the Top 25 Colleges</h3>

In [101]:
fig = px.bar(top25, x='college', y='points_per_game', hover_data = ["count"], color='college', 
                 title = "Points Per Game for the Top 25 Colleges")
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

<h3 style="color:blue; font-weight:bold">Rebounds Per Game for the Top 25 Colleges</h3>

In [102]:
fig = px.bar(top25, x='college', y='average_total_rebounds', hover_data = ["count"], color='college', 
                 title = "Rebounds Per Game for the Top 25 Colleges")
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

<h3 style="color:blue; font-weight:bold">Assists Per Game for the Top 25 Colleges</h3>

In [103]:
fig = px.bar(top25, x='college', y='average_assists', hover_data = ["count"], color='college', 
                 title = "Assists Per Game for the Top 25 Colleges")
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

<h3 style="color:blue; font-weight:bold">Box Plus Minus for the Top 25 Colleges</h3>

In [104]:
fig = px.bar(top25, x='college', y='box_plus_minus', hover_data = ["count"], color='college', 
                 title = "Box Plus Minus for the Top 25 Colleges")
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

<h3 style="color:blue; font-weight:bold">Value Over Replacement for the Top 25 Colleges</h3>

In [105]:
fig = px.bar(top25, x='college', y='value_over_replacement', hover_data = ["count"], color='college', 
                 title = "Value Over Replacement for the Top 25 Colleges")
fig.update_layout(template = "plotly_dark", font = dict(family = "PT Sans", size = 12))
fig.show()

<h2 style="color:red; font-weight:bold">Success of Draft Picks</h2>
<h3 style="color:blue; font-weight:bold">Create a method for valuing each draft slot in the NBA Draft</h3>

I devised a method to evaluate the success of a draft pick by comparing a player's performance to the historical average of players chosen at the same draft position. This involved calculating average statistics across all relevant categories in the NBA draft dataset, grouping them by overall pick position. By comparing a player's stats to these averages, I generated a count of how many statistical categories the player exceeded the average. If a player surpassed the average in 9 or more statistical categories, they were classified as a "hit." As a disclamer, these stats do not reflect seasons after the 2021 meaning some players maybe considered a miss if playing was not high in their rookie/sophomore year.

Let's begin by identifying the draft positions that have yielded the most successful players based on their statistics, and then we'll rank them accordingly.

In [106]:
df.fillna(0, inplace=True)
# drop columns that are not needed
draft = df.drop(['id', 'year', 'rank', 'player', 'team', 'college'], axis=1)
# group all rows by overall pick value and get the mean for each row
draft = draft.groupby('overall_pick').mean().reset_index()
c_names = draft.columns.tolist()
c_names.pop(0)
# go through each column and create a new column that will contain it's rank for that statistic
for col in c_names:
    name = col + "_rank"
    draft[name] = draft[col].rank(ascending=False)
draft = draft.drop(columns = c_names)
# create the picks final rank by getting the average of all statistical rankings together
draft['rank']= draft.mean(axis=1)
temp = draft.sort_values(by=['rank'])
rank_list = temp["overall_pick"].values.tolist()
print("Rank: ")
print(*rank_list, sep = ", ") 

Rank: 
3, 4, 1, 5, 2, 7, 10, 9, 13, 8, 6, 11, 12, 17, 14, 24, 15, 21, 18, 23, 27, 16, 19, 26, 22, 20, 25, 28, 30, 29, 35, 45, 37, 31, 38, 40, 32, 42, 34, 33, 36, 43, 47, 41, 46, 39, 48, 56, 44, 57, 49, 50, 55, 60, 52, 51, 54, 58, 53, 59


<h3 style="color:blue; font-weight:bold">How many teams hit on their Picks</h3>

In [107]:
draft_avg = df.drop(['id', 'year', 'rank', 'player', 'team', 'college'], axis=1)
draft_avg = draft_avg.groupby('overall_pick').mean().reset_index()
team = df['team'].unique()
team = sorted(team)
temp = {
    'team': ['ATL','BOS','BRK','CHI','CHO','CLE','DAL','DEN',
 'DET','GSW','HOU','IND','LAC','LAL','MEM','MIA','MIL','MIN',
 'NOP','NYK','OKC','ORL','PHI','PHO','POR','SAC','SAS','TOR','UTA','WAS'],
 'hits': [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
 'total': [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
}
team_success = pd.DataFrame(temp)

### Categories used for comparison:
* 'years_active'
* 'games'
* 'minutes_played'
* 'points'
* 'total_rebounds'
* 'assists'
* 'field_goal_percentage'
* '3_point_percentage' 
* 'free_throw_percentage'
* 'average_minutes_played'
* 'points_per_game'
* 'average_total_rebounds'
* 'average_assists' 
* 'win_shares'
* 'win_shares_per_48_minutes'
* 'box_plus_minus'
* 'value_over_replacement'

If a players is has 9 statistical categories that are better than the average value, he will be considered a "hit". 




In [108]:
# for every team go through each pick and compare the player's stats to the pick average
for t in team:
    temp = df[df['team'] == t]
    # total_pick will be used to keep track of the total number of picks for the given team
    total_picks = 0
    # count will be used to keep track of the "hits"
    count = 0
    for index, player in temp.iterrows():
        avg_pick = player['overall_pick']
        avg_row = draft_avg[draft_avg['overall_pick'] == avg_pick]
        total = 0
        for c in c_names:
            a = float(player[c])
            b = float(avg_row[c])
            if (a >= b):
                total += 1
        total_picks +=1
        if total >= 9:
            count += 1
    # update the team's values after evaluationg each pick in the team_success df
    team_success.loc[team_success['team'] == t, 'hits'] = count
    team_success.loc[team_success['team'] == t, 'total'] = total_picks

# create a new column in the team_success df, it wull be the percentage of picks that are considered hits
team_success['pct_hit'] = team_success['hits']/team_success['total']

In [109]:
cols = ['pct_hit']
team_success.style.background_gradient(cmap='seismic',subset=cols).hide(axis="index")


The San Antonio Spurs are the most successful franchise when it comes to drafting players, while the Portland Trail Blazers are the least successful in this regard.